cmd/virtual-kubelet/internal/commands/root/root.go

```diff
@@ -142,119 +131,129 @@ func runRootCommand(ctx context.Context, s *provider.Store, c Opts) error {
+			Workers:                c.PodSyncWorkers,
+			PodInformerFactory:     podInformerFactory,
+			GenericInformerFactory: scmInformerFactory,
+		},
+	})
+
 	go podInformerFactory.Start(ctx.Done())
 	go scmInformerFactory.Start(ctx.Done())
 
-	cancelHTTP, err := setupHTTPServer(ctx, p, apiConfig, func(context.Context) ([]*corev1.Pod, error) {
-		return rm.GetPods(), nil
-	})
-	if err != nil {
-		return err
+	go m.Run(ctx)
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-m.Ready():
+	case <-m.Done():
+		return m.Err()
 	}
-	defer cancelHTTP()
 
-	go func() {
-		if err := pc.Run(ctx, c.PodSyncWorkers); err != nil && errors.Cause(err) != context.Canceled {
-			log.G(ctx).Fatal(err)
-		}
-	}()
-
-	if c.StartupTimeout > 0 {
-		ctx, cancel := context.WithTimeout(ctx, c.StartupTimeout)
-		log.G(ctx).Info("Waiting for pod controller / VK to be ready")
-		select {
-		case <-ctx.Done():
-			cancel()
-			return ctx.Err()
-		case <-pc.Ready():
-		}
-		cancel()
-		if err := pc.Err(); err != nil {
-			return err
+	if pm, ok := p.(provider.PodMetricsProvider); ok {
+		stop, err := setupMetricsServer(c, pm)
+		if err != nil {
+			return errors.Wrap(err, "error setting up pod metrics server")
 		}
+		defer stop()
 	}
 
-	go func() {
-		if err := nodeRunner.Run(ctx); err != nil {
-			log.G(ctx).Fatal(err)
-		}
-	}()
+	setNodeReady(pNode)
+	if err := np.UpdateStatus(ctx, pNode); err != nil {
+		return err
+	}
 
 	log.G(ctx).Info("Initialized")
 
-	<-ctx.Done()
+	select {
+	case <-ctx.Done():
+	case <-m.Done():
+		return m.Err()
+	}
 	return nil
 }
 
-func newClient(configPath string) (*kubernetes.Clientset, error) {
-	var config *rest.Config
-
-	// Check if the kubeConfig file exists.
-	if _, err := os.Stat(configPath); !os.IsNotExist(err) {
-		// Get the kubeconfig from the filepath.
-		config, err = clientcmd.BuildConfigFromFlags("", configPath)
-		if err != nil {
-			return nil, errors.Wrap(err, "error building client config")
+func setNodeReady(n *corev1.Node) {
+	for i, c := range n.Status.Conditions {
+		if c.Type != "Ready" {
+			continue
 		}
-	} else {
-		// Set to in-cluster config.
-		config, err = rest.InClusterConfig()
-		if err != nil {
-			return nil, errors.Wrap(err, "error building in cluster config")
-		}
-	}
 
-	if masterURI := os.Getenv("MASTER_URI"); masterURI != "" {
-		config.Host = masterURI
+		c.Message = "Kubelet is ready"
+		c.Reason = "KubeletReady"
+		c.Status = corev1.ConditionTrue
+		c.LastHeartbeatTime = metav1.Now()
+		c.LastTransitionTime = metav1.Now()
+		n.Status.Conditions[i] = c
+		return
 	}
 
-	return kubernetes.NewForConfig(config)
+	// No ready condition in  node status
+	c := corev1.NodeCondition{
+		Type:               "Ready",
+		Status:             corev1.ConditionTrue,
+		Reason:             "KubeletReady",
+		Message:            "Kubelet is ready",
+		LastHeartbeatTime:  metav1.Now(),
+		LastTransitionTime: metav1.Now(),
+	}
+	n.Status.Conditions = append(n.Status.Conditions, c)
 }
```

Comment:
Not related to this change, but the code duplication between node-cli and cmd/virtual-kubelet/internal makes me wonder if node-cli really should be a separate repo.

Reply:
The reason for the split is API guarantees.
`node-cli` does not have a stable API, but we wanted to have a way to help people prototype providers, remove some boiler plate, etc.

This PR is really an attempt to bring some of that work into a stable API.